File: scripts/extract.py
#!/usr/bin/env python3

import sys
import json
import pytz
from datetime import datetime, timezone

def convert_timestamp(unix_timestamp, from_tz, to_tz):
    """
    Convert a UNIX timestamp from one timezone to another.
    """
    # Convert UNIX timestamp to a datetime object in the 'from' timezone
    dt = datetime.fromtimestamp(int(unix_timestamp), from_tz)

    # Convert the datetime object to the 'to' timezone
    dt_to = dt.astimezone(to_tz)

    # Return the UNIX timestamp of the converted datetime
    return int(dt_to.timestamp())

def main():
    if len(sys.argv) < 2:
        print("Usage: extract.py <EVENT_ID> <OUTCOME_ID>")
        sys.exit(1)

    event_id = sys.argv[1]
    outcome_id = sys.argv[2]
    output_filename = f"./generated/{event_id}-{outcome_id}.tsv"

    # Define timezones (adjust as needed)
    from_timezone = pytz.UTC  # Assuming the original timestamps are in UTC
    to_timezone = pytz.timezone('America/New_York')  # Example: converting to Eastern Time

    with open(f"output-{event_id}.tsv", "r") as infile, open(output_filename, "w") as outfile:
        # Write the header to the CSV
        outfile.write("timestamp\timplied_probability\tamerican\thandicap\n")

        # Iterate through each line of the input
        for line in infile:
            parts = line.strip().split('\t')
            if "PriceEvent" in line and outcome_id in line:
                try:
                    # Convert timestamp from one timezone to another
                    converted_timestamp = convert_timestamp(parts[1], from_timezone, to_timezone)

                    # Try to extract the decimal price from the JSON string
                    event = json.loads(parts[3])

                    # Default handicap value
                    handicap = 0

                    # Update handicap if it exists in the data
                    if "handicap" in event["price"]:
                        handicap = float(event["price"]["handicap"])

                    decimal_price = float(event["price"]["decimal"])
                    american_price = float(event["price"]["american"])
                    implied_probability = round(1 / decimal_price, 2)

                    # Write the processed data to the output file
                    outfile.write(f"{converted_timestamp}\t{implied_probability}\t{american_price}\t{handicap}\n")
                except (IndexError, ValueError, KeyError, TypeError) as e:
                    # Handling potential JSON parsing errors, or missing keys
                    continue

if __name__ == "__main__":
    main()

File: scripts/extract.js
#!/usr/bin/env node

const fs = require('fs')

const round = (value, decimals) => {
    return Number(Math.round(value + 'e' + decimals) + 'e-' + decimals)
} 

const main = async () => {
    if (process.argv.length < 4) {
        console.error('Usage: extract.py <EVENT_ID> <OUTCOME_ID>')
        process.exit(1)
    }
    const eventId = process.argv[2]
    const outcomeId = process.argv[3]
    const outputFilename = `./generated/${eventId}-${outcomeId}.tsv`
    const input = await fs.promises.readFile(`./output-${eventId}.tsv`, 'utf8')
    const lines = input.split('\n')
    let output = 'timestamp\timplied_probability\tamerican\thandicap\n'
    let numRows = 0
    for (const line of lines) {
        if (!line.includes('PriceEvent')) {
            continue
        }
        if (!line.includes(outcomeId)) {
            continue
        }
        try {
            const parts = line.trim().split('\t')
            const timestamp = parts[1]
            const event = JSON.parse(parts[3])
            let handicap = 0
            if (event.price.handicap) {
                handicap = parseFloat(event.price.handicap)
            }
            const decimalPrice = parseFloat(event.price.decimal)
            const americanPrice = parseFloat(event.price.american)
            const impliedProbability = round(1 / decimalPrice, 2)
            output += `${timestamp}\t${impliedProbability}\t${americanPrice}\t${handicap}\n`
            numRows += 1
        } catch (err) {
            console.error(err)
        }
    }
    if (numRows === 0) {
        throw new Error('numRows === 0')
    }
    await fs.promises.writeFile(outputFilename, output)
}

main()
